src/saltext/vmware/states/esxi.py

```diff
@@ -1,3 +1,216 @@
-"""
-Nothing is here yet, this is a placeholder to be able to build docs.
-"""
+# SPDX-License-Identifier: Apache-2.0
+import logging
+
+from saltext.vmware.utils.connect import get_service_instance
+
+
+log = logging.getLogger(__name__)
+
+try:
+    import pyVmomi
+
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+
+__virtualname__ = "vmware_esxi"
+__proxyenabled__ = ["vmware_esxi"]
+
+
+def __virtual__():
+    if not HAS_PYVMOMI:
+        return False, "Unable to import pyVmomi module."
+    return __virtualname__
+
+
+def user_present(
+    name,
+    password,
+    description=None,
+    datacenter_name=None,
+    cluster_name=None,
+    host_name=None,
+    service_instance=None,
+):
+    """
+    Add local user on matching ESXi hosts.
+
+    name
+        User to create on matching ESXi hosts. (required).
+
+    password
+        Password for the user. (required).
+
+    description
+        Description of the user. (optional).
+
+    datacenter_name
+        Filter by this datacenter name (required when cluster is specified)
+
+    cluster_name
+        Filter by this cluster name (optional)
+
+    host_name
+        Filter by this ESXi hostname (optional)
+
+    service_instance
+        Use this vCenter service connection instance instead of creating a new one. (optional).
+
+    .. code-block:: yaml
+
+    Create User:
+      vmware_esxi.user_present:
+        - name: local_user
+        - password: secret
+
+    """
+    log.debug("Running vmware_esxi.user_present")
+    ret = {"name": name, "result": None, "comment": "", "changes": {}}
+    create = update = 0
+    diff = {}
+    if not service_instance:
+        service_instance = get_service_instance(opts=__opts__, pillar=__pillar__)
+    user = __salt__["vmware_esxi.get_user"](
+        user_name=name,
+        datacenter_name=datacenter_name,
+        cluster_name=cluster_name,
+        host_name=host_name,
+        service_instance=service_instance,
+    )
+    hosts = __salt__["vmware_esxi.list_hosts"](
+        datacenter_name=datacenter_name,
+        cluster_name=cluster_name,
+        host_name=host_name,
+        service_instance=service_instance,
+    )
+    for host in hosts:
+        if host in user:
+            # Cannot determine if password is changed. So, when the user is found on the host, update it.
+            diff[host] = {
+                "old": {"name": name, "description": user[host][name]["description"]},
+                "new": {"name": name, "description": description},
+                "action": "update",
+            }
+            update += 1
+        else:
+            diff[host] = {"new": {"name": name, "description": description}, "action": "create"}
+            create += 1
+    for host in diff:
```

Comment:
It seems like `update_user` updates multiple hosts, so no iteration is required here.

Reply:
it is a mixed bag as some of the users need to be created and some others need to be updated on certain hosts. Hence, iterating to create and update users one by one, as appropriate, filtering by host.

Reply:
I guess that does make sense. In that case, the `ret` is clobbered on every loop; and only the result from the last execution is returned. That doesn't seem right.

Reply:
I thought about it, will think more if there is a way to update it only once after the end of the iteration instead of everytime.

Reply:
done

Reply:
I like the message. 

My only remaining concern is that when running against a large set of hosts, any error will terminate the execution.

Let's say you have 2000 hosts to update but at any time some are down for maintenance. Assuming the list is ordered, you will essentially never be able to update the last machine in the list; if the list were randomized you would eventually be successful.

I think the better approach is to accumulate a list of hosts that had errors, then raise an exception with text like this:

"User {} created on {} host(s) and updated on {} host(s). Errors occurred for {} hosts; sample error: {}"? 

"{} hosts" could be either an integer or a list of hosts (or both)? so: "Errors occurred for {} hosts: {}\n sample error: {}"?

Reply:
ok will change